### ISSUE TITLE
Add ability to specify 'extra' resources

### ISSUE THREAD

**Issue by ThatJoeMoore** [byu-oit/handel-codepipeline#60]
Somewhat related to #38 .

It would be nice to have a section in the pipeline configuration that lets us specify a subset of the resources available in Handel that our pipeline needs to run, and have it set those up and configure the proper permissions.

This differs from having a 'handel' phase, as this is for resources that are needed by the build process itself, not for the application.  Theoretically, this could be accomplished by having a 'handel' phase with a 'build' environment, but we cannot currently have a 'handel' phase before our 'codebuild' phase, which is where we would most likely need these resources.

Here's what I propose:

```
...
pipelines:
  my_pipeline:
    phases:
      - ...
    extra_resources:
      - <service_name>:
         type: <service_type>
        <service_param>: <param_value>
        dependencies:
          - <service name>
```

# Example Use Case

In the lovely Java world in which my team works, from-scratch builds can take a very long time, due to long dependency resolution times.  In environments like CircleCI, we can use a build cache to reduce these dependency resolution times dramatically, since our dependency tree is rather static.

To implement a build cache, we would probably have a step in our 'pre_build' CodeBuild step that goes out to an S3 bucket and downloads the latest copy of the cache.  Once the build is done, we could bundle up the local cache and stick it back out on S3.

While we could manually set up the S3 bucket and configure its permissions, it would be nice (and less error-prone) to have Handel do it for us.

**Comment by ThatJoeMoore** [comment]
Note that, with some guidance, I would be happy to be the one to implement this.

**Comment by dsw88** [comment]
We talked about just supporting S3 initially as an attached resource in the CodeBuild phase, but keeping it generic enough so that we could add more extra resources when needed. 

Since the resources are probably only needed in the CodeBuild phase, what you do think about this amended schema?
```
...
pipelines:
  my_pipeline:
    phases:
    ...
    - type: codebuild
      ...
      extra_resources:
      - type: <service_type>
         <service_param>: <param_value>
```

Note that I've proposed moving the 'extra_resources' section into the configuration for the codebuild phase type itself, since that's about the only phase type in the pipeline that gives you the capability to execute arbitrary commands. I've also proposed not supporting 'dependencies' to start, since things like S3 or Dynamo that we're likely to support won't have complex dependency hierarchies. If I'm ever proven wrong we can add 'dependencies' here since it would just be an additive change.

What do you think about those proposed changes?

**Comment by ThatJoeMoore** [comment]
LGTM

**Comment by ThatJoeMoore** [comment]
I'm finally starting on this.  @dsw88 , what do you think of changing `extra_resources` to be a map, so that it more closely resembles a normal Handel file?  This way, we can easily name the resources, instead of having to diverge from Handel and put the name as a property in the resource object.

So, like this:

```yaml
...
pipelines:
  my_pipeline:
    phases:
    ...
    - type: codebuild
      ...
      extra_resources:
        build-cache:
          type: <service_type>
          <service_param>: <param_value>
```

**Comment by dsw88** [comment]
I think that change to the schema is a good idea, let's do it. I do want to talk to you about this change before you get started on it because I had an idea that might simplify its implementation, especially if we ever had a need to have other service types in addition to S3.

**Comment by ThatJoeMoore** [comment]
Sounds good. I'm gone tomorrow, but I'll talk to you on Monday. So far, I'm
just writing documentation for how it would work - my usual flow is to
write docs, then tests, then code.

**Comment by dsw88** [comment]
I've moved this issue into the [roadmap](https://github.com/byu-oit/handel-codepipeline/projects/1) for implementation.

**Comment by dsw88** [comment]
Resolved by https://github.com/byu-oit/handel-codepipeline/pull/66
